feat(stellar): add link token support (#853)
axelarnetwork/axelar-contract-deployments: stellar/docs/link-token.md

## How It Works

The token linking process involves two key message types that ITS Hub uses to coordinate the linking:

### Token Metadata Registration

Before linking tokens, ITS Hub needs to know about each token's details (address and decimals) on both chains.

### Token Linking

Once metadata is registered, you can link tokens by specifying which tokens on different chains should be connected.

### Decimal Scaling

ITS Hub automatically handles decimal scaling between linked tokens:

- **Stellar Token**: 7 decimals (1 USDC = 10,000,000 units)
- **EVM Token**: 18 decimals (1 USDC = 1,000,000,000,000,000,000 units)
- **Scaling Factor**: 10^(18-7) = 10^11

When transferring 1 USDC from Stellar to EVM:

- Stellar: 10,000,000 units locked/burned
- EVM: 1,000,000,000,000,000,000 units unlocked/minted

## Link Token Flow

```
1. User controls Token A (Stellar) and Token B (EVM)
2. User → ITS Stellar: registerTokenMetadata(Token A) → ITS Hub
3. User → ITS EVM: registerTokenMetadata(Token B) → ITS Hub
4. User → ITS Stellar: registerCustomToken() → Deploys Token Manager A
5. User → ITS Stellar: linkToken() → ITS Hub
6. ITS Hub: Calculates scaling factor from stored decimals

> 7. ITS Hub → ITS EVM: Deploy Token Manager B

Reviewer: What ITS hub does is a design detail, whereas the flow should cover what the user needs to do to link the token

8. Token linking complete - InterchainTransfer enabled
```

## Prerequisites

Before linking tokens, ensure you have:

1. **Token Control**: You must control both tokens on their respective chains

> 2. **Token Metadata**: Both tokens must be registered with their metadata

Reviewer: this is not a prerequisite, but part of the instructions below

3. **Gas Tokens**: Sufficient gas tokens for cross-chain operations
4. **ITS Deployment**: ITS contracts must be deployed on both chains

## Token Manager Types

The following token manager types are supported:

- `LOCK_UNLOCK` (2): For tokens that are locked on the source chain and unlocked on the destination chain
- `MINT_BURN` (4): For tokens that are burned on the source chain and minted on the destination chain

## Parameters

**Required:**

- `salt`: Unique identifier for the token linking operation
- `tokenAddress`: Address of the token to be linked
- `destinationChain`: Name of the destination chain
- `destinationTokenAddress`: Address of the token on the destination chain
- `type`: Token manager type (LOCK_UNLOCK, MINT_BURN)

**Optional:**

- `--operator`: Operator address for the token manager on the destination chain
- `--gas-token-address`: Gas token address (defaults to XLM on Stellar)
- `--gas-amount`: Gas amount for cross-chain operations

## Step-by-Step Process

### Step 1: Setup Tokens

Setup Token with decimals on both chains:

**Chain A (Stellar):**

```bash

> ts-node stellar/its.js deploy-interchain-token <name> <symbol> <decimal> <salt> <initialSupply>

Reviewer: These are test tokens. We are assuming the user already has existing tokens, so this needs to be labelled correctly

```

**Chain B (EVM):**

```bash
ts-node evm/interchainTokenFactory.js \
  --action deployInterchainToken \
  --minter <minterAddress> \
  --name <name> \
  --symbol <symbol>
  --decimals <decimal>
  --initialSupply <initialSupply>
  --salt <salt>
  -n <network>
```

### Step 2: Register Token Metadata

**Chain A (Stellar):**

```bash
ts-node stellar/its.js register-token-metadata <tokenAddress> --gas-amount <gasAmount>
```

**Chain B (EVM):**

```bash
ts-node evm/its.js --action registerTokenMetadata \
  --tokenAddress <tokenAddress>
  -n <network>
```

### Step 3: Register Custom Token

Register the token on the source chain (Stellar):

```bash
ts-node stellar/its.js register-custom-token <salt> <tokenAddress> <tokenManagerType>
```

### Step 4: Link Token

Link the token to the destination chain:

```bash
ts-node stellar/its.js link-token <salt> <destinationChain> <destinationTokenAddress> <tokenManagerType> \
   --gas-amount <amount> \
   --operator <operator>

> ```

Reviewer: missing steps on transferring mint/burn rights depending on token manager type

## Examples

> ### Example 1: Link USDC Tokens with LOCK_UNLOCK

Reviewer: ```suggestion
### Example 1: Link Ethereum USDC with your own USDC contract on Stellar
```

Link USDC tokens with different decimals (7 decimals on Stellar, 18 decimals on EVM):

```bash
# USDC already exists on both chains

# Register USDC metadata on Stellar (7 decimals)
ts-node stellar/its.js register-token-metadata CB64D3G...USDC --gas-amount 10000000

# Register USDC metadata on EVM (18 decimals)
ts-node evm/its.js --action registerTokenMetadata --tokenAddress 0xa0b86a33...USDC -n evm_chain

# Register custom token on Stellar
ts-node stellar/its.js register-custom-token 0x1234 CB64D3G...USDC LOCK_UNLOCK

# Link token to EVM
ts-node stellar/its.js link-token 0x1234 evm_chain 0xa0b86a33...USDC LOCK_UNLOCK --gas-amount 10000000

# Interchain Token Transfer
ts-node stellar/its.js interchain-transfer <tokenId> evm_chain <destinationAddress> <amount> --gas-amount 10000000
```

> ### Example 2: Link Custom Token with MINT_BURN

Reviewer: not a very useful example, since most cmd args are stubbed out. one example is sufficient

```bash
# Register custom token metadata on both chains
ts-node stellar/its.js register-token-metadata <stellarTokenAddress> --gas-amount 10000000
ts-node evm/its.js --action registerTokenMetadata --tokenAddress <ethereumTokenAddress> -n evm_chain

# Register and link using MINT_BURN type
ts-node stellar/its.js register-custom-token <salt> <stellarTokenAddress> MINT_BURN
ts-node stellar/its.js link-token <salt> evm_chain <ethereumTokenAddress> MINT_BURN --gas-amount 10000000 --operator <operatorAddress>

# Get token Manager on EVM
ts-node evm/its.js --action tokenManagerAddress --tokenId <tokenId> -n evm_chain

# Transfer mintership on EVM
ts-node evm/its.js --action transferMintership --tokenAddress <tokenAddress> --minter <tokenManager> -n evm_chain

# Interchain Token Transfer
ts-node stellar/its.js interchain-transfer <tokenId> evm_chain <destinationAddress> <amount> --gas-amount 10000000
```

## Troubleshooting & Error Handling

**Invalid Token Manager Type:**

```
Error: Invalid token manager type: INVALID_TYPE. Valid types: LOCK_UNLOCK, MINT_BURN
```

Solution: Use a valid token manager type from the supported list.

**Token Not Registered:**

```

> Error: Token metadata not found in ITS Hub

Reviewer: add a cosmwasm query before submitting `linkToken` to ensure token metadata is registered on ITS Hub. The script should throw readable errors such that this section isn't even needed

Author: I'm not sure it could be directly queried to get [custom tokens state](https://github.com/axelarnetwork/axelar-amplifier/blob/d0ba17c71aefb8a254fa06583fd0ec6b11821fda/contracts/interchain-token-service/src/state.rs#L166); could you help me how to find the query command?

```

Solution: Register token metadata on both chains before linking.

**Insufficient Gas:**

```
Error: Insufficient gas for cross-chain operation
```

Solution: Increase the gas amount using `--gas-amount` option.

**Token Already Linked:**

```
Error: Token already linked with this salt
```

Solution: Use a different salt value for the linking operation.

## Best Practices & Security

> 1. **Salt Management**: Use unique, cryptographically secure salts for each token linking operation

Reviewer: incorrect, salt can be any string, it just needs to be unique per token id being linked

2. **Gas Estimation**: Always estimate gas costs before executing operations
3. **Token Verification**: Verify token addresses and decimals before linking
4. **Token Control**: Ensure you have proper control over both tokens

> 5. **Operator Security**: Use secure operator addresses with appropriate permissions

Author: updated

6. **Transaction Verification**: Check gas amount and token permissions if transactions fail

> 7. **Chain Connectivity**: Ensure both chains are connected to ITS Hub

Author: removed

# Link Token Documentation

## Overview

This document explains how to link custom tokens on Stellar to other chains using the Interchain Token Service (ITS).

For detailed design specifications and architecture, see **ARC-1: ITS Hub Multi-Chain Token Linking**:

> https://github.com/axelarnetwork/arcs/blob/main/arcs/arc-1.md

Reviewer 2: We can use permalink in this case

> https://github.com/axelarnetwork/arcs/blob/main/arcs/arc-1.md

Author: Thank you for the comment, updated

The token linking feature enables:

- Linking custom tokens deployed on Stellar with tokens on EVM chains
- Supporting tokens with different decimal precisions through automatic scaling

## How It Works

The token linking process involves two key message types that ITS Hub uses to coordinate the linking:

### Token Metadata Registration

Before linking tokens, ITS Hub needs to know about each token's details (address and decimals) on both chains.

### Token Linking

Once metadata is registered, you can link tokens by specifying which tokens on different chains should be connected.

## Link Token Flow

```
1. User controls Token A (Stellar) and Token B (EVM)
2. User → ITS Stellar: registerTokenMetadata(Token A)
3. User → ITS EVM: registerTokenMetadata(Token B)
4. User → ITS Stellar: registerCustomToken() → Deploys Token Manager A

> 5. User → Verify token metadata is registered on ITS Hub

Reviewer 3: Can we specify command for doing this? Or perhaps link to this section where this command is explained

6. User → ITS Stellar: linkToken() → Deploys Token Manager B on destination chain
7. User → Transfer or add mintership to the Token Manager (MINT_BURN type only)
8. Token linking complete - InterchainTransfer enabled
```

## Prerequisites

Before linking tokens, ensure you have:

1. **Token Control**: Token permissions depend on the manager type:
    - **MINT_BURN**: You must have minter permissions for the token on that chain

>     - **LOCK_UNLOCK**: No token control or minter permissions required for that token

Reviewer 3: I dont think permission is required, but I believe their needs to be an `allowance` set for this manager type as it calls `transferFrom()` on the token, which I think is worth calling out.

    **Note:** Since you cannot link two LOCK_UNLOCK tokens, one token must be MINT_BURN (requiring minter permissions) and the other can be LOCK_UNLOCK (no permissions required).

## Token Manager Types

The following token manager types are supported:

- `LOCK_UNLOCK` (2): For tokens that are locked on the source chain and unlocked on the destination chain
- `MINT_BURN` (4): For tokens that are burned on the source chain and minted on the destination chain

> **Important:** You cannot have two LOCK_UNLOCK tokens linked together. At most one token can be LOCK_UNLOCK type. The other token must be MINT_BURN type and owned by the issuer with mint rights given to the token manager.

Reviewer 3: Perhaps elaborate on the why? Technically I think devs can still deploy multiple lock/unlock. It's just highly recommended they dont do that as managing supply on tokens is very hard/dangerous, and there is no `withdraw()` fn to get your token out if you choose to go that route

Author: updated

## Parameters

**Required:**

> - `salt`: Unique identifier for the token linking operation

Reviewer 3: ```suggestion
- `salt`: Unique identifier for the token linking operation. Used to generate a unique `tokenId`
```

- `tokenAddress`: Address of the token to be linked
- `destinationChain`: Name of the destination chain
- `destinationTokenAddress`: Address of the token on the destination chain
- `type`: Token manager type (e.g. LOCK_UNLOCK, MINT_BURN)

## Operator Role & Security

The `--operator` parameter specifies an address that controls the token manager on the destination chain.

**Operator can:**

- Set and modify flow limits
- Pause/unpause token manager operations

**Security:** The operator cannot steal tokens directly, but can modify settings that affect interchain token service. Use trusted addresses only.

> **Note:** The deployer account (caller of linkToken) must also be secure, as it has the authority to initiate token linking operations.

Reviewer 3: ```suggestion
**Note:** The deployer account (caller of `linkToken`) must also be secure, as it has the authority to initiate token linking operations.
```

## Step-by-Step Process

### Step 1: Setup Tokens

**Note: This step is for deploying test tokens. If you want to use existing tokens, skip this step and proceed to Step 2.**

Deploy Test Tokens on both chains:

**Chain A (Stellar):**

```bash
ts-node stellar/its.js deploy-interchain-token <name> <symbol> <decimal> <salt> <initialSupply>
```

**Alternative - Create a Stellar classic asset:**

```bash

> # Optional trust limit (defaults to 1000000000 if not specified)

Reviewer 3: what is this trust limit? I dont think we explained it anywhere

ts-node stellar/token-utils.js create-stellar-classic-asset [asset-code] [issuer] [limit]
```

**Chain B (EVM):**

```bash
ts-node evm/interchainTokenFactory.js \
  --action deployInterchainToken \
  --minter <minterAddress> \
  --name <name> \
  --symbol <symbol>
  --decimals <decimal>
  --initialSupply <initialSupply>
  --salt <salt>
  -n <network>
```

### Step 2: Register Token Metadata

#### For Stellar Classic Assets

> If you're linking a Stellar Classic asset (format: {Symbol-Issuer}) that doesn't have a Soroban contract address yet, you can deploy a corresponding Soroban contract to make them accessible within Soroban-based contracts:

Reviewer 3: ```suggestion
If you're linking a Stellar Classic asset (format: {Symbol-Issuer}) that doesn't have a Soroban contract address yet, you can deploy a corresponding Stellar contract to make them accessible within Stellar-based contracts:
```
I think I recall we're avoiding using "soroban" as a term?
soroban

```
1. User controls Token A (Stellar) and Token B (EVM)
2. User → ITS Stellar: registerTokenMetadata(Token A)
3. User → ITS EVM: registerTokenMetadata(Token B)
4. User → ITS Stellar: registerCustomToken() → Deploys Token Manager A
5. User → Verify token metadata is registered on ITS Hub
6. User → ITS Stellar: linkToken() → Deploys Token Manager B on destination chain
7. User → Transfer or add mintership to the Token Manager (MINT_BURN and MINT_BURN_FROM type only)
8. Token linking complete - InterchainTransfer enabled
```

## Prerequisites

Before linking tokens, ensure you have:

- **Token Control**: Token permissions depend on the manager type:
    - **LOCK_UNLOCK**: No token control or minter permissions required for that token
    - **MINT_BURN_FROM / MINT_BURN**: You must have permission to transfer or add mintership for the token on that chain

## Token Manager Types

The following token manager types are supported:

- `MINT_BURN_FROM` (1): For tokens that are burned/minted using burnFrom/mint functions on the chain
- `LOCK_UNLOCK` (2): For tokens that are locked/unlocked on the chain
- `MINT_BURN` (4): For tokens that are burned/minted on the chain

> **Important:** Linking two LOCK_UNLOCK tokens is highly not recommended. One token should be MINT_BURN or MINT_BURN_FROM (requiring minter permissions) and the other can be LOCK_UNLOCK (no permissions required).

Reviewer 4: ```suggestion
**Important:** Linking two LOCK_UNLOCK tokens is highly discouraged. One token should be MINT_BURN or MINT_BURN_FROM (requiring minter permissions) and the other can be LOCK_UNLOCK (no permissions required).
```

or 

```suggestion
**Important:** Linking two LOCK_UNLOCK tokens is not recommended. One token should be MINT_BURN or MINT_BURN_FROM (requiring minter permissions) and the other can be LOCK_UNLOCK (no permissions required).
```

```

**On EVM:**

```bash
# Get token manager address on the destination chain
ts-node evm/its --action tokenManagerAddress --tokenId <tokenId> -n <destinationChain>

# Transfer mintership to the token manager
ts-node evm/its --action transferMintership --tokenAddress <tokenAddress> --minter <tokenManagerAddress> -n <destinationChain>
```

### Optional: Transfer Token Admin (MINT_BURN Type Only)

**Note: This step is optional and only applicable when using the MINT_BURN token manager type. For MINT_BURN_FROM and LOCK_UNLOCK types, skip this step.**

For `MINT_BURN` token managers, once you transfer admin to the token manager, you need to request ITS owner to transfer the admin back to you.

**On Stellar:**

```bash
# Transfer token admin to the new admin address (only ITS Owner is allowed to call transfer-token-admin)
ts-node stellar/token-utils transfer-token-admin <tokenId> <newAdminAddress>
```

## Examples

### Example 1: Link Ethereum USDC (LOCK_UNLOCK) with a Stellar Classic Asset (MINT_BURN)

Link USDC tokens with different decimals (19 decimals on EVM, 7 decimals on Stellar):

```bash
# Register USDC metadata on EVM (18 decimals)

> ts-node evm/its --action registerTokenMetadata --tokenAddress 0xa0b86a33...USDC -n evm_chain

Reviewer 4: ```suggestion
ts-node evm/its --action registerTokenMetadata --tokenAddress 0xa0b86a33...USDC -n <evm_chain>

Link USDC tokens with different decimals (18 decimals on EVM, 7 decimals on Stellar):

```bash
# Register USDC metadata on EVM
ts-node evm/its --action registerTokenMetadata --tokenAddress 0xa0b86a33...USDC -n evm_chain

# Register custom token metadata on Stellar
# Custom token should exist at this point
ts-node stellar/its register-token-metadata CB64D3G...TEST --gas-amount 10000000

# Verify token metadata is registered on ITS Hub
ts-node cosmwasm/query token-config stellar CB64D3G...TEST
ts-node cosmwasm/query token-config evm_chain 0xa0b86a33...USDC

# Register custom token on Stellar (MINT_BURN_FROM)
ts-node stellar/its register-custom-token 0x1234 CB64D3G...TEST MINT_BURN_FROM

# Link token to EVM (LOCK_UNLOCK type for the existing Ethereum USDC)
ts-node stellar/its link-token 0x1234 evm_chain 0xa0b86a33...TEST LOCK_UNLOCK --gas-amount 10000000 --operator <operatorAddress>
# Result: 89a0c5...abcd (token id)

# Get token manager address
ts-node stellar/its deployed-token-manager 0x89a0c5...abcd
# Result: CATERX...ABCD (token manager address)

# Add minter permission based on your token implementation

# Interchain Token Transfer

> ts-node stellar/its interchain-transfer 0x89a0c5...abcd evm_chain 0xba76...dest 1 --gas-amount 10000000

Reviewer 4: A bit hard to read the parameters
```suggestion
ts-node stellar/its interchain-transfer <0x89a0c5...abcd> <evm_chain> <0xba76...dest> 1 --gas-amount 10000000
```
perhaps?

### Step 2: Register Token Metadata

#### For Stellar Classic Assets

If you're using Stellar Classic assets (format: {Symbol-Issuer}), you must deploy a corresponding Soroban contract to make them accessible within Soroban-based contracts:

**Stellar Classic Asset Types:**

- Classic assets follow the {Symbol-Issuer} format

> - Addresses starting with "C" indicate Soroban smart contracts
> - Classic assets can be deployed as Soroban contracts for ITS compatibility

Reviewer: ```suggestion
```
not really needed

> **Note:** If you already have a Soroban contract (address starting with "C"), you can skip the deploy step below and register token metadata directly.

Reviewer: not needed now
```suggestion
```

> **Deploy Soroban Contract for Classic Asset:**

Reviewer: ```suggestion
```
already discussed above

```bash
ts-node stellar/token-utils.js create-stellar-asset-contract <assetCode> <issuer>
```

Example:

```bash
# For USDC Classic asset (USDC-GA5ZSEJYB37JRC5AVCIA5MOP4RHTM335X2KGX3IHOJAPP5RE34K4KZVN)
ts-node stellar/token-utils.js create-stellar-asset-contract USDC GA5ZSEJYB37JRC5AVCIA5MOP4RHTM335X2KGX3IHOJAPP5RE34K4KZVN

# Result: CCW67TSZV3SSS2HXMBQ5JFGCKJNXKZM7UQUWUZPUTHXSTZLEO7SJMI75
```

> **Note:** If you're using the Axelar ITS Portal UI, it will automatically call this function to deploy the Soroban contract.
> 

Reviewer: ```suggestion
```
user is using the script to deploy if they're reading this

ts-node stellar/its.js register-custom-token <salt> <tokenAddress> <tokenManagerType>
```

### Step 4: Verify Token Metadata Registration

Before linking tokens, verify that token metadata is registered on ITS Hub for both chains:

```bash
# Verify source token metadata (Stellar)
axelard query wasm contract-state smart $ITS_HUB_ADDRESS '{"custom_tokens":{"chain":"<sourceChain>","token_address":"<tokenAddress>"}}' --output json

# Verify destination token metadata (EVM)
axelard query wasm contract-state smart $ITS_HUB_ADDRESS '{"custom_tokens":{"chain":"<destinationChain>","token_address":"<destinationTokenAddress>"}}' --output json

> ```

Reviewer: do this in the script, instead of making the user perform it. you can grep a simple example of how to do a cosmwasm query in the scripts

If either query fails or returns null, ensure you complete Step 2 (Register Token Metadata) before proceeding.

### Step 5: Link Token

Link the token to the destination chain:

- `--operator`: Operator address for the token manager on the destination chain

```bash
ts-node stellar/its.js link-token <salt> <destinationChain> <destinationTokenAddress> <tokenManagerType> \
   --gas-amount <amount> \
   --operator <operatorAddress>
```

### Step 6: Transfer Minter Permissions (MINT_BURN Type Only)

**Note: This step is only required if you're using the MINT_BURN token manager type. Skip this step for LOCK_UNLOCK type.**

For MINT_BURN token managers, you must transfer minter permissions to the token manager on both chains:

> **Get Token Manager Address:**

Reviewer: ```suggestion
```

```bash

> # Get token manager address on the destination chain

Reviewer: move this to on evm section instead of moving back and forth between chains

ts-node evm/its.js --action tokenManagerAddress --tokenId <tokenId> -n <destinationChain>
```

**Transfer Minter Permissions:**

**On Stellar:**

```bash
# Transfer minter permissions to token manager (if applicable)

> # This depends on your token's specific implementation

Reviewer: add a transfer method for our interchain token? we won't be able to test ourselves without this

Author: updated

```

**On EVM:**

> 

Reviewer: update the private key in .env to EVM wallet
